servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2ResponseCancelTest.java

```diff
@@ -0,0 +1,188 @@
+
+    @Test
+    public void testClient() throws Exception {
+        // Release existing connection to avoid creating unexpected number of new h2 connections because of randomness
+        // in RRLB (it may randomly pick the reserved connection until all attempts are taken).
+        ((ReservedStreamingHttpConnection) streamingHttpConnection()).releaseAsync().toFuture().get();
+
+        int connectionsBefore = newConnectionsCounter.get();
+        requestCancellationResetsStreamButNotParentConnection(streamingHttpClient());
+        int connectionsAfter = newConnectionsCounter.get();
+        assertThat("Client unexpectedly created more connections instead of reusing the existing one",
+                connectionsAfter - connectionsBefore, is(0));
+    }
+
+    @Test
+    public void testConnection() throws Exception {
+        StreamingHttpConnection connection = streamingHttpConnection();
+        AtomicBoolean connectionClosed = new AtomicBoolean();
+        connection.onClose().whenFinally(() -> connectionClosed.set(true)).subscribe();
+
+        requestCancellationResetsStreamButNotParentConnection(connection);
+        assertThat("Connection closed unexpectedly", connectionClosed.get(), is(false));
+    }
+
+    @Test
+    public void testServerClosesStreamNotConnection() throws Exception {
+        StreamingHttpConnection connection = streamingHttpConnection();
+        AtomicBoolean connectionClosed = new AtomicBoolean();
+        connection.onClose().whenFinally(() -> connectionClosed.set(true)).subscribe();
+
+        ExecutionException e = assertThrows(ExecutionException.class,
+                () -> makeRequest(newRequest(connection, "close")));
+        assertThat(e.getCause(), instanceOf(H2StreamResetException.class));
+
+        // Mak sure we can use the same connection for future requests:
+        assertResponse(makeRequest(newRequest(connection, "ok")), HTTP_2_0, OK, "ok");
+        assertThat("Connection closed unexpectedly", connectionClosed.get(), is(false));
+    }
+
+    private void requestCancellationResetsStreamButNotParentConnection(StreamingHttpRequester requester)
+            throws Exception {
+
+        BlockingQueue<StreamingHttpResponse> responses = new LinkedBlockingDeque<>();
+        requester.request(defaultStrategy(), newRequest(requester, "first")).subscribe(responses::add);
+        firstRequestReceivedLatch.await();
+
+        AtomicReference<Cancellable> cancellable = new AtomicReference<>();
+        stepVerifier(requester.request(defaultStrategy(), newRequest(requester, "second"))
+                        .whenOnSuccess(responses::add)) // Add response to the queue to verify that we never receive it
+                .expectCancellableConsumed(cancellable::set)
+                .then(() -> {
+                    try {
+                        secondRequestReceivedLatch.await();
+                        cancellable.get().cancel(); // If I use thenCancel() the current then(Runnable) does not run
```

Comment:
@Scottmitch if I do this way:
```java
stepVerifier(requester.request(defaultStrategy(), newRequest(requester, "second"))
                .whenOnSuccess(responses::add)) // Add response to the queue to verify that we never receive it
        .expectCancellable()
        .then(() -> {
            try {
                secondRequestReceivedLatch.await();
            } catch (InterruptedException e) {
                // ignore
            }
        })
        .thenCancel()
        .verify();
```
`then(Runnable)` is not invoked, after `expectCancellable()` the next step will be `thenCancel()`. Is this expected behavior?

Reply:
discussed offline, we will investigate clarifying/improving as part of followup issues. tldr; the "then" runnable operations are run on the verifier thread, the expectations are run on the subscriber thread (including `thenCancel`), so these aren't sequential.